clinica/iotools/converters/ukb_to_bids/ukb_utils.py

```diff
@@ -0,0 +1,319 @@
+from cmath import nan
+from os import PathLike
+from typing import BinaryIO, Iterable, List, Tuple, Union
+
+from pandas import DataFrame, notna
+
+
+def find_clinical_data(
+    clinical_data_directory: PathLike,
+) -> DataFrame:
+    from pathlib import Path
+
+    import pandas as pd
+
+    from clinica.utils.stream import cprint
+
+    # Read the xls
+    try:
+        image_data_file = list(Path(clinical_data_directory).glob("*.tsv"))
+    except StopIteration:
+        raise FileNotFoundError("Clinical data file not found.")
+
+    # Assign the dfs
+    if len(image_data_file) == 1:
+        df_clinical = pd.read_csv(str(image_data_file[0]), sep="\t")
+        cprint(msg="All clinical data have been found", lvl="info")
+    elif len(image_data_file) == 0:
+        raise FileNotFoundError("Clinical data not found or incomplete")
+    elif len(image_data_file) > 1:
+        raise FileNotFoundError("Too many data files found")
+    return df_clinical
+
+
+def read_imaging_data(imaging_data_directory: PathLike) -> DataFrame:
+    from pathlib import Path
+
+    import pandas as pd
+
+    source_path_series = pd.Series(
+        find_imaging_data(imaging_data_directory), name="source_path"
+    )
+    # list of the files we want to build the bids for each modality
+    file_mod_list = [
+        "T1_orig_defaced.nii.gz",
+        "T2_FLAIR_orig_defaced.nii.gz",
+        "AP.nii.gz",
+    ]
+
+    dataframe = pd.DataFrame.from_records(
+        source_path_series, columns=["source_zipfile", "source_filename"]
+    )
+    filename = (
+        dataframe["source_filename"]
+        .apply(lambda x: Path(str(x)).name)
+        .rename("filename")
+    )
+    dataframe = dataframe[filename.isin(file_mod_list)]
+    filename = filename[filename.isin(file_mod_list)]
+    split_zipfile = dataframe["source_zipfile"].str.split("_", expand=True)
+    split_zipfile = split_zipfile.rename(
+        {0: "Subject", 1: "modality_num", 2: "session_number"}, axis="columns"
+    ).drop_duplicates()
+    split_zipfile = split_zipfile.astype({"Subject": "int64"})
+
+    df_source = pd.concat(
+        [
+            dataframe,
+            filename,
+            split_zipfile,
+        ],
+        axis=1,
+    )
+    return df_source
+
+
+def find_imaging_data(path_to_source_data: PathLike) -> Iterable[PathLike]:
+    from pathlib import Path
+    from zipfile import ZipFile
+
+    for z in Path(path_to_source_data).rglob("*.zip"):
+        for f in ZipFile(z).namelist():
+            if f.endswith(".nii.gz"):
+                yield [str(z.relative_to(path_to_source_data)), f]
+
+
+def intersect_data(
+    df_source: DataFrame, df_clinical_data: DataFrame
+) -> Tuple[DataFrame, DataFrame]:
```

Comment:
This function currently does two different things:

- It performs a `merge` on `df_clinical_data` and `df_source` and assign it to `df_clinical` 
- Does some DataFrame operations on `df_clinical` without needing anything else
- And it returns `df_source` without doing any modifications on it 

It looks like you could split this in at least two functions:

- `intersect_source_and_clinical_data(df_source, df_clinical_data) -> df_clinical`
- `clean_clinical_data(df_clinical) -> df_clinical`

Or something like this, WDYT?

Reply:
Sounds like a sensible proposal. Having a cleaning / sanitization process for each dataframe, followed by a merge step.

Reply:
Sounds good, I'll look into it right now 😄

Reply:
@ghisvail @NicolasGensollen I have changed the intersect_data function and split in two. Is it what you were thinking of ?

Reply:
This is how a *somewhat* generic converter looks like to me:

```mermaid
flowchart LR
LCD --> CCD --> MCI
LID --> CID --> MCI
MCI --> PCI --> WCI
```

with:
- LCD, LID: Load clinical / imaging data
- CCD, CID: Clean clinical / imaging data
- MCI: Merge clinical and imaging data
- PCI: Prepare merged data for writing
- WCI: Write merged data to BIDS

If we all agree with this, then we could keep those names.

Reply:
I have been thinking about this for a while (since we talked about migrating to `pydra`), and having a generic structure for a converter could help standardizing towards a workflow compatible design.

Reply:
@ghisvail I agree. If all converters follow (or could potentially be adapted to follow) this pattern, then it would definitely be beneficial to standardize them.

Reply:
Okay. I won't change this right now since other modification might change our point of view on this, but i will come back to you both when all the modalities are included in the converter.